File: day11/python/main.py
import math
import copy

class Monkey:
    items: list[int]
    operation: [str, int]
    test: int
    throw: [int, int]
    inspection: int

    def __init__(self) -> None:
        self.items = []
        self.operation = []
        self.test = 0
        self.throw = [-1, -1]
        self.inspection = 0

    def __repr__(self) -> str:
        return f"Monkey {{ items: {self.items}, operation: {self.operation}, test: {self.test}, throw: {self.throw}, inspection: {self.inspection}}}"


def parse_input(lines: list[str]) -> list[Monkey]:
    monkeys = []
    for line in lines:
        if line[0] == "\n":
            continue
        if line[0] == "M":
            monkeys.append(Monkey())
            continue
        if line[2] == "S":
            monkey = monkeys[-1]
            items = line.split(":")[1]
            items = items.split(",")
            items = [int(item) for item in items]
            monkey.items = items
            continue
        if line[2] == "O":
            monkey = monkeys[-1]
            line = line[line.find("=") + 1:]
            split = line.split("+")
            if len(split) == 2:
                monkey.operation.append("add")
            else:
                split = line.split("*")
                monkey.operation.append("mul")

            if split[1].find("old") == -1:
                monkey.operation.append(int(split[1]))
            else:
                monkey.operation.append(-1)
            continue
        if line[2] == "T":
            monkey = monkeys[-1]
            monkey.test = int(line.split("by")[1])
            continue
        if line[4] == "I":
            monkey = monkeys[-1]
            value = int(line.split("monkey")[-1])
            if line.find("true") != -1:
                monkey.throw = [value, monkey.throw[1]]
            else:
                monkey.throw = [monkey.throw[0], value]
            continue
    return monkeys


def first_task(monkeys: list[Monkey]) -> int:
    for _ in range(20):
        for monkey in monkeys:
            while len(monkey.items) != 0:
                item = monkey.items.pop(0)
                monkey.inspection += 1
                worry = 0
                operation_value = item if monkey.operation[1] == - \
                    1 else monkey.operation[1]
                if monkey.operation[0] == "add":
                    worry = item + operation_value
                else:
                    worry = item * operation_value
                worry = int(worry / 3)
                test = worry % monkey.test
                if test == 0:
                    monkeys[monkey.throw[0]].items.append(worry)
                else:
                    monkeys[monkey.throw[1]].items.append(worry)

    inspections = [monkey.inspection for monkey in monkeys]
    inspections.sort()
    return inspections[-1] * inspections[-2]


def second_task(monkeys: list[Monkey]) -> int:
    tests = []
    for monkey in monkeys:
        tests.append(monkey.test)
    test_lcm = math.lcm(*tests)
    for _ in range(10000):
        for monkey in monkeys:
            while len(monkey.items) != 0:
                item = monkey.items.pop(0)
                monkey.inspection += 1
                worry = 0
                operation_value = item
                if monkey.operation[1] != -1:
                    operation_value = monkey.operation[1]
                if monkey.operation[0] == "add":
                    worry = item + operation_value
                else:
                    worry = item * operation_value

                if worry % monkey.test == 0:
                    monkeys[monkey.throw[0]].items.append(worry)
                else:
                    monkeys[monkey.throw[1]].items.append(worry % test_lcm)
    
    inspections = [monkey.inspection for monkey in monkeys]
    inspections.sort()
    return inspections[-1] * inspections[-2]


def main():
    file = open("input.txt", "r")
    lines = file.readlines()
    monkeys = parse_input(lines)
    first_task_result = first_task(copy.deepcopy(monkeys))
    second_task_result = second_task(monkeys)
    print(f"first_task: {first_task_result}")
    print(f"second_task: {second_task_result}")
    # The asserts are added after the solution to ensure refactoring does not brake it.
    assert(first_task_result == 110220)
    assert(second_task_result == 19457438264)

if __name__ == "__main__":
    main()

File: day11/rust/src/main.rs
use num::integer::lcm;
use std::fs;

#[derive(Debug, Clone)]
struct Monkey {
    items: Vec<i64>,
    operation: (String, i64),
    test: i64,
    throw: (usize, usize),
    inspection: i64,
}

fn parse_input(lines: Vec<&str>) -> Vec<Monkey> {
    let mut monkeys = Vec::new();
    for line in lines {
        if line.is_empty() {
            continue;
        }
        let mut chars = line.chars();
        let char0 = chars.next().unwrap();
        if char0 == 'M' {
            monkeys.push(Monkey {
                items: Vec::new(),
                operation: ("".to_string(), 0),
                test: 0,
                throw: (0, 0),
                inspection: 0,
            });
            continue;
        }

        let char3 = chars.nth(1).unwrap();
        if char3 == 'S' {
            let monkey = monkeys.last_mut().unwrap();
            let items: Vec<i64> = line[line.find(':').unwrap() + 1..]
                .split(',')
                .map(|x| x[1..].parse::<i64>().unwrap())
                .collect();
            monkey.items = items;
        }
        if char3 == 'O' {
            let monkey = monkeys.last_mut().unwrap();
            let after_equal = line[line.find('=').unwrap() + 1..].to_string();
            let split_add = after_equal.split('+').collect::<Vec<&str>>();
            let split_mul = after_equal.split('*').collect::<Vec<&str>>();
            if split_add.len() == 2 {
                monkey.operation.0 = "add".to_string();
                monkey.operation.1 = if split_add[1].contains("old") {
                    -1
                } else {
                    split_add[1][1..].parse::<i64>().unwrap()
                };
            } else {
                monkey.operation.0 = "mul".to_string();
                monkey.operation.1 = if split_mul[1].contains("old") {
                    -1
                } else {
                    split_mul[1][1..].parse::<i64>().unwrap()
                };
            }
            continue;
        }

        if char3 == 'T' {
            let monkey = monkeys.last_mut().unwrap();
            monkey.test = line[line.find("by ").unwrap() + "by ".len()..]
                .parse::<i64>()
                .unwrap();
            continue;
        }

        let char5 = chars.nth(1).unwrap();
        if char5 == 'I' {
            // It must be I anyway but still...
            let monkey = monkeys.last_mut().unwrap();
            let value = line[line.find("monkey ").unwrap() + "monkey ".len()..]
                .parse::<usize>()
                .unwrap();
            if line.contains("true") {
                monkey.throw.0 = value;
            } else {
                monkey.throw.1 = value;
            }
        }
    }

    monkeys
}

fn first_task(mut monkeys: Vec<Monkey>) -> i64 {
    for _ in 0..20 {
        for i in 0..monkeys.len() {
            while !monkeys[i].items.is_empty() {
                let item = monkeys[i].items[0];
                monkeys[i].items.remove(0);
                monkeys[i].inspection += 1;
                let operation_value = if monkeys[i].operation.1 == -1 {
                    item
                } else {
                    monkeys[i].operation.1
                };
                let worry = if monkeys[i].operation.0 == "add" {
                    (item + operation_value) / 3
                } else {
                    (item * operation_value) / 3
                };
                if worry % monkeys[i].test == 0 {
                    let index = monkeys[i].throw.0;
                    monkeys[index].items.push(worry);
                } else {
                    let index = monkeys[i].throw.1;
                    monkeys[index].items.push(worry);
                }
            }
        }
    }
    let mut inspections = monkeys.iter().map(|x| x.inspection).collect::<Vec<i64>>();
    inspections.sort();
    inspections[inspections.len() - 1] * inspections[inspections.len() - 2]
}

fn second_task(mut monkeys: Vec<Monkey>) -> i64 {
    let mut test_lcm: i64 = 1;
    for monkey in monkeys.iter() {
        test_lcm = lcm(test_lcm, monkey.test);
    }
    for _count in 0..10000 {
        for i in 0..monkeys.len() {
            while !monkeys[i].items.is_empty() {
                let item = monkeys[i].items[0];
                monkeys[i].items.remove(0);
                monkeys[i].inspection += 1;
                let operation_value = if monkeys[i].operation.1 == -1 {
                    item
                } else {
                    monkeys[i].operation.1
                };
                let worry = if monkeys[i].operation.0 == "add" {
                    item + operation_value
                } else {
                    item * operation_value
                };
                if worry % monkeys[i].test == 0 {
                    let index = monkeys[i].throw.0;
                    monkeys[index].items.push(worry);
                } else {
                    let index = monkeys[i].throw.1;
                    monkeys[index].items.push(worry % test_lcm);
                }
            }
        }
    }
    let mut inspections = monkeys.iter().map(|x| x.inspection).collect::<Vec<i64>>();
    inspections.sort();
    inspections[inspections.len() - 1] * inspections[inspections.len() - 2]
}

fn main() {
    let file_content = fs::read_to_string("../input.txt").unwrap();
    let monkeys = parse_input(file_content.lines().collect());
    let first_task_result = first_task(monkeys.clone());
    let second_task_result = second_task(monkeys);
    println!("first_task: {first_task_result}");
    println!("second_task: {second_task_result}");
    // Assertion values are obtained from output of python script.
    assert_eq!(first_task_result, 110220);
    assert_eq!(second_task_result, 19457438264);
}
